{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import time\n",
    "import random\n",
    "import requests\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "from typing import List, Union\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import f_oneway\n",
    "\n",
    "from pysradb import SRAweb\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "from IPython.display import SVG\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = Path.cwd().parent.parent.parent / 'HIV_Atlas_Creation' / 'data'\n",
    "base_dir = Path.cwd().parent.parent\n",
    "soft_dir = base_dir / 'soft'\n",
    "\n",
    "prj_id = 'BoltonSIV'\n",
    "\n",
    "outdir = base_dir / 'results' / prj_id\n",
    "outdir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "prj_aln_dir = outdir / 'alignment'\n",
    "prj_aln_dir.mkdir(parents=True, exist_ok=True)\n",
    "prj_assembly_dir = outdir / 'assembly'\n",
    "prj_assembly_dir.mkdir(parents=True, exist_ok=True)\n",
    "prj_quant_dir = outdir / 'quantification'\n",
    "prj_quant_dir.mkdir(parents=True, exist_ok=True)\n",
    "prj_aux_dir = outdir / 'auxiliary'\n",
    "prj_aux_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "hisat_idx_basename = base_dir / 'data/SIV239/reference'\n",
    "reference_fasta_fname = base_dir / 'data/SIV239/reference.fasta'\n",
    "reference_gtf_fname = base_dir / 'data/SIV239/reference.gtf'\n",
    "\n",
    "sivPlus_info_fname = data_dir / \"info_SIVplus.tsv\"\n",
    "\n",
    "sashimi_bin = \"sashimi.py\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 1\n",
    "\n",
    "sys.path.insert(0, str(soft_dir / \"genomic_scripts\"))\n",
    "%aimport definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{' Memory CD4 T cell gene expression from macaque AY69 lymph node', ' Memory CD4 T cell gene expression from macaque T034 PBMC'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>animal</th>\n",
       "      <th>sample</th>\n",
       "      <th>Cell number</th>\n",
       "      <th>SIV reads</th>\n",
       "      <th>SIV prop</th>\n",
       "      <th>SIV prop &gt;0.01</th>\n",
       "      <th>in final filtered seurat</th>\n",
       "      <th>D1-A5 splice</th>\n",
       "      <th>tatrev splice</th>\n",
       "      <th>SIV pos</th>\n",
       "      <th>number of splice varieties seen</th>\n",
       "      <th>passed initial sc QC</th>\n",
       "      <th>Experiment Accession</th>\n",
       "      <th>Experiment Title</th>\n",
       "      <th>Sample Accession</th>\n",
       "      <th>Total Spots</th>\n",
       "      <th>Total Bases</th>\n",
       "      <th>Run</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AY69</td>\n",
       "      <td>A1p1</td>\n",
       "      <td>1</td>\n",
       "      <td>42031</td>\n",
       "      <td>0.020604</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>127</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>SRX20459058</td>\n",
       "      <td>GSM7399872: Memory CD4 T cell gene expression ...</td>\n",
       "      <td>SRS17766270</td>\n",
       "      <td>2585301</td>\n",
       "      <td>661808315</td>\n",
       "      <td>SRR24679672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AY69</td>\n",
       "      <td>A2p1</td>\n",
       "      <td>1</td>\n",
       "      <td>43091</td>\n",
       "      <td>0.022958</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>264</td>\n",
       "      <td>841</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>SRX20458082</td>\n",
       "      <td>GSM7399879: Memory CD4 T cell gene expression ...</td>\n",
       "      <td>SRS17765295</td>\n",
       "      <td>2683173</td>\n",
       "      <td>660469952</td>\n",
       "      <td>SRR24678661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AY69</td>\n",
       "      <td>A6p1</td>\n",
       "      <td>1</td>\n",
       "      <td>157834</td>\n",
       "      <td>0.072799</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2250</td>\n",
       "      <td>3347</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>SRX20458792</td>\n",
       "      <td>GSM7399907: Memory CD4 T cell gene expression ...</td>\n",
       "      <td>SRS17766006</td>\n",
       "      <td>2702252</td>\n",
       "      <td>664149793</td>\n",
       "      <td>SRR24679087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AY69</td>\n",
       "      <td>A8p1</td>\n",
       "      <td>1</td>\n",
       "      <td>123108</td>\n",
       "      <td>0.104491</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1159</td>\n",
       "      <td>1089</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>SRX20458806</td>\n",
       "      <td>GSM7399921: Memory CD4 T cell gene expression ...</td>\n",
       "      <td>SRS17766018</td>\n",
       "      <td>1741331</td>\n",
       "      <td>447735007</td>\n",
       "      <td>SRR24679073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AY69</td>\n",
       "      <td>B12p1</td>\n",
       "      <td>1</td>\n",
       "      <td>111957</td>\n",
       "      <td>0.073472</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>961</td>\n",
       "      <td>1613</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>SRX20458844</td>\n",
       "      <td>GSM7399949: Memory CD4 T cell gene expression ...</td>\n",
       "      <td>SRS17766056</td>\n",
       "      <td>2032111</td>\n",
       "      <td>518745652</td>\n",
       "      <td>SRR24679461</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  animal sample  Cell number  SIV reads  SIV prop  SIV prop >0.01  \\\n",
       "0   AY69   A1p1            1      42031  0.020604               1   \n",
       "1   AY69   A2p1            1      43091  0.022958               1   \n",
       "2   AY69   A6p1            1     157834  0.072799               1   \n",
       "3   AY69   A8p1            1     123108  0.104491               1   \n",
       "4   AY69  B12p1            1     111957  0.073472               1   \n",
       "\n",
       "   in final filtered seurat  D1-A5 splice  tatrev splice  SIV pos  \\\n",
       "0                         1           127              0        1   \n",
       "1                         1           264            841        1   \n",
       "2                         1          2250           3347        1   \n",
       "3                         1          1159           1089        1   \n",
       "4                         1           961           1613        1   \n",
       "\n",
       "   number of splice varieties seen  passed initial sc QC Experiment Accession  \\\n",
       "0                                1                     1          SRX20459058   \n",
       "1                                2                     1          SRX20458082   \n",
       "2                                2                     1          SRX20458792   \n",
       "3                                2                     1          SRX20458806   \n",
       "4                                2                     1          SRX20458844   \n",
       "\n",
       "                                    Experiment Title Sample Accession  \\\n",
       "0  GSM7399872: Memory CD4 T cell gene expression ...      SRS17766270   \n",
       "1  GSM7399879: Memory CD4 T cell gene expression ...      SRS17765295   \n",
       "2  GSM7399907: Memory CD4 T cell gene expression ...      SRS17766006   \n",
       "3  GSM7399921: Memory CD4 T cell gene expression ...      SRS17766018   \n",
       "4  GSM7399949: Memory CD4 T cell gene expression ...      SRS17766056   \n",
       "\n",
       "   Total Spots  Total Bases          Run  \n",
       "0      2585301    661808315  SRR24679672  \n",
       "1      2683173    660469952  SRR24678661  \n",
       "2      2702252    664149793  SRR24679087  \n",
       "3      1741331    447735007  SRR24679073  \n",
       "4      2032111    518745652  SRR24679461  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_df = pd.read_csv(sivPlus_info_fname, sep='\\t', index_col=0).reset_index()\n",
    "print(set(metadata_df[\"Experiment Title\"].str.split(\":\",expand=True)[1].str.split(\";\",expand=True)[0].str.split(\",\",expand=True)[0].unique()))\n",
    "metadata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# align samples with hisat\n",
    "cmd_lst_fname = prj_aln_dir / 'cmd_lst.txt'\n",
    "reads_dir = data_dir / \"data/fastq/\"\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"Run\"].tolist():\n",
    "        r1_fname = reads_dir+run_id+\"_1.fastq\"\n",
    "        r2_fname = reads_dir+run_id+\"_2.fastq\"\n",
    "        cmd = f\"hisat2 -p 64 --score-min L,0,-2 --mp 2,2 -x {hisat_idx_basename} -1 {r1_fname} -2 {r2_fname} -S {prj_aln_dir}/{run_id}.sam\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sort reads by position\n",
    "cmd_lst_fname = prj_aln_dir / 'cmd_lst.sort.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"Run\"].tolist():\n",
    "        cmd = f\"samtools sort -@ 64 -o {prj_aln_dir}/{run_id}.sorted.bam {prj_aln_dir}/{run_id}.sam\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run tiebrush to generate joint representations of each monkey\n",
    "\n",
    "# for each monkey write out a list of alignments in a .lst file\n",
    "for monkey in metadata_df[\"animal\"].unique():\n",
    "    lst_fname = prj_aln_dir / f'{monkey}.lst'\n",
    "    with open(lst_fname, 'w') as outFP:\n",
    "        for run_id in metadata_df[metadata_df[\"animal\"]==monkey][\"Run\"].tolist():\n",
    "            outFP.write(f\"{prj_aln_dir}/{run_id}.sorted.bam\\n\")\n",
    "\n",
    "cmd_lst_fname = prj_aln_dir / 'cmd_lst.tiebrush.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for monkey in metadata_df[\"animal\"].unique():\n",
    "        lst_fname = prj_aln_dir / f'{monkey}.lst'\n",
    "        cmd = f\"tiebrush -o {prj_aln_dir}/{monkey}.tb.bam {lst_fname}\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract SIV reads only (M33262.1)\n",
    "cmd_lst_fname = prj_aln_dir / 'cmd_lst.extract.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for monkey in metadata_df[\"animal\"].unique():\n",
    "        cmd = f\"samtools index {prj_aln_dir}/{monkey}.tb.bam && samtools view -h {prj_aln_dir}/{monkey}.tb.bam M33262.1 | samtools sort -o {prj_aln_dir}/{monkey}.tb.M33262.1.sorted.bam -\"\n",
    "        outFP.write(f\"{cmd}\\n\")\n",
    "        for run_id in metadata_df[metadata_df[\"animal\"]==monkey][\"Run\"].tolist():\n",
    "            cmd = f\"samtools index {prj_aln_dir}/{run_id}.sorted.bam && samtools view -h {prj_aln_dir}/{run_id}.sorted.bam M33262.1 | samtools sort -o {prj_aln_dir}/{run_id}.M33262.1.sorted.bam -\"\n",
    "            outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run tiecov\n",
    "cmd_lst_fname = prj_aln_dir / 'cmd_lst.tiecov.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for monkey in metadata_df[\"animal\"].unique():\n",
    "        cmd = f\"tiecov -c {prj_aln_dir}/{monkey}.tb.M33262.1.coverage.bedgraph -j {prj_aln_dir}/{monkey}.tb.M33262.1.junctions.bed {prj_aln_dir}/{monkey}.tb.M33262.1.sorted.bam\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assemble with stringtie\n",
    "cmd_lst_fname = prj_assembly_dir / 'cmd_lst.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"Run\"].tolist():\n",
    "        cmd = f\"stringtie -p 64 -G {reference_gtf_fname} -o {prj_assembly_dir}/{run_id}.gtf {prj_aln_dir}/{run_id}.sorted.bam\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge stringtie results to identify and quantify novel transcripts\n",
    "monkey_cmp_lst = [] # results of gffcompare for each monkey to aggregate into a single gtf for downstream analysis as well\n",
    "for monkey in metadata_df[\"animal\"].unique():\n",
    "    lst_fname = prj_assembly_dir / f'{monkey}.lst'\n",
    "    with open(lst_fname, 'w') as out_lstFP:\n",
    "        for run_id in metadata_df[metadata_df[\"animal\"]==monkey][\"Run\"].tolist():\n",
    "            out_lstFP.write(f\"{prj_assembly_dir}/{run_id}.gtf\\n\")\n",
    "    definitions.run_gffcompare({\"-r\": reference_gtf_fname, \"-p\": f\"{monkey}\", \"-o\": f\"{prj_assembly_dir}/{monkey}\", \"-i\": lst_fname})\n",
    "    monkey_cmp_lst.append(f\"{prj_assembly_dir}/{monkey}.combined.gtf\")\n",
    "\n",
    "# merge all gtf files\n",
    "cmd = f\"gffcompare -r {reference_gtf_fname} -p ALL -o {prj_assembly_dir}/ALL {' '.join(monkey_cmp_lst)}\"\n",
    "print(cmd)\n",
    "subprocess.run(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create subsets of the gffcompare output for the M33262.1 genome only\n",
    "for monkey in metadata_df[\"animal\"].unique():\n",
    "    gtf_fname = prj_assembly_dir / f'{monkey}.combined.gtf'\n",
    "    tracking_fname = prj_assembly_dir / f'{monkey}.tracking'\n",
    "    \n",
    "    # load the tids with the the seqid\n",
    "    gtf_df = definitions.get_attribute(gtf_fname, [\"transcript_id\"],[0])\n",
    "    tids = gtf_df[gtf_df[0]==\"M33262.1\"][\"transcript_id\"].tolist()\n",
    "    \n",
    "    # subset to the M33262.1 genome\n",
    "    # subset gtf\n",
    "    sub_gtf_fname = prj_assembly_dir / f'{monkey}.M33262.1.gtf'\n",
    "    sub_tracking_fname = prj_assembly_dir / f'{monkey}.M33262.1.tracking'\n",
    "    definitions.subset_gtf(gtf_fname,sub_gtf_fname,[],tids)\n",
    "    definitions.subset_tracking(tracking_fname,sub_tracking_fname,tids)\n",
    "    \n",
    "# subset ALL tracking and gtf\n",
    "gtf_fname = prj_assembly_dir / f'ALL.combined.gtf'\n",
    "tracking_fname = prj_assembly_dir / f'ALL.tracking'\n",
    "# load the tids with the the seqid\n",
    "gtf_df = definitions.get_attribute(gtf_fname, [\"transcript_id\"],[0])\n",
    "tids = gtf_df[gtf_df[0]==\"M33262.1\"][\"transcript_id\"].tolist()\n",
    "# subset gtf\n",
    "sub_gtf_fname = prj_assembly_dir / f'ALL.M33262.1.gtf'\n",
    "sub_tracking_fname = prj_assembly_dir / f'ALL.M33262.1.tracking'\n",
    "definitions.subset_gtf(gtf_fname,sub_gtf_fname,[],tids)\n",
    "definitions.subset_tracking(tracking_fname,sub_tracking_fname,tids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run orfanage on novel transcripts\n",
    "cmd = f\"orfanage --query {prj_assembly_dir}/ALL.M33262.1.gtf --reference {reference_fasta_fname} --output {prj_assembly_dir}/ALL.M33262.1.orfanage.gtf {reference_gtf_fname}\"\n",
    "print(cmd)\n",
    "subprocess.run(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pull tracking data into the gtf\n",
    "# initialize the tracking hierarchy\n",
    "all_tracking_fname = prj_assembly_dir / 'ALL.M33262.1.tracking'\n",
    "all_name = \"ALL\"\n",
    "hierarchy = {\n",
    "    (all_name,all_tracking_fname):{}\n",
    "}\n",
    "for monkey in metadata_df[\"animal\"].unique():\n",
    "    mnk_tracking_fname = prj_assembly_dir / f'{monkey}.M33262.1.tracking'\n",
    "    hierarchy[(all_name,all_tracking_fname)][(monkey,mnk_tracking_fname)] = {}\n",
    "\n",
    "definitions.combine_tracking_gtf(prj_assembly_dir / 'ALL.M33262.1.orfanage.gtf', hierarchy, prj_assembly_dir / 'ALL.M33262.1.orfanage.tracking.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>class_code</th>\n",
       "      <th>ALL_tpm_mean</th>\n",
       "      <th>ALL_num_samples</th>\n",
       "      <th>AY69_tpm_mean</th>\n",
       "      <th>AY69_num_samples</th>\n",
       "      <th>T034_tpm_mean</th>\n",
       "      <th>T034_num_samples</th>\n",
       "      <th>ALL_percent_samples</th>\n",
       "      <th>AY69_percent_samples</th>\n",
       "      <th>T034_percent_samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>ALL_00411239</td>\n",
       "      <td>=</td>\n",
       "      <td>982.157097</td>\n",
       "      <td>149.0</td>\n",
       "      <td>1183.173731</td>\n",
       "      <td>89.0</td>\n",
       "      <td>683.982423</td>\n",
       "      <td>60.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>ALL_00411254</td>\n",
       "      <td>=</td>\n",
       "      <td>371.188454</td>\n",
       "      <td>127.0</td>\n",
       "      <td>365.079713</td>\n",
       "      <td>85.0</td>\n",
       "      <td>383.551384</td>\n",
       "      <td>42.0</td>\n",
       "      <td>0.852349</td>\n",
       "      <td>0.955056</td>\n",
       "      <td>0.700000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>ALL_00411255</td>\n",
       "      <td>=</td>\n",
       "      <td>239.708965</td>\n",
       "      <td>126.0</td>\n",
       "      <td>262.067336</td>\n",
       "      <td>85.0</td>\n",
       "      <td>193.356246</td>\n",
       "      <td>41.0</td>\n",
       "      <td>0.845638</td>\n",
       "      <td>0.955056</td>\n",
       "      <td>0.683333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ALL_00411242</td>\n",
       "      <td>=</td>\n",
       "      <td>289.746285</td>\n",
       "      <td>125.0</td>\n",
       "      <td>214.829095</td>\n",
       "      <td>86.0</td>\n",
       "      <td>454.948295</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.838926</td>\n",
       "      <td>0.966292</td>\n",
       "      <td>0.650000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>ALL_00411243</td>\n",
       "      <td>=</td>\n",
       "      <td>138.904371</td>\n",
       "      <td>125.0</td>\n",
       "      <td>125.493695</td>\n",
       "      <td>86.0</td>\n",
       "      <td>168.476630</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.838926</td>\n",
       "      <td>0.966292</td>\n",
       "      <td>0.650000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             tid class_code  ALL_tpm_mean  ALL_num_samples  AY69_tpm_mean  \\\n",
       "69  ALL_00411239          =    982.157097            149.0    1183.173731   \n",
       "72  ALL_00411254          =    371.188454            127.0     365.079713   \n",
       "38  ALL_00411255          =    239.708965            126.0     262.067336   \n",
       "44  ALL_00411242          =    289.746285            125.0     214.829095   \n",
       "53  ALL_00411243          =    138.904371            125.0     125.493695   \n",
       "\n",
       "    AY69_num_samples  T034_tpm_mean  T034_num_samples  ALL_percent_samples  \\\n",
       "69              89.0     683.982423              60.0             1.000000   \n",
       "72              85.0     383.551384              42.0             0.852349   \n",
       "38              85.0     193.356246              41.0             0.845638   \n",
       "44              86.0     454.948295              39.0             0.838926   \n",
       "53              86.0     168.476630              39.0             0.838926   \n",
       "\n",
       "    AY69_percent_samples  T034_percent_samples  \n",
       "69              1.000000              1.000000  \n",
       "72              0.955056              0.700000  \n",
       "38              0.955056              0.683333  \n",
       "44              0.966292              0.650000  \n",
       "53              0.966292              0.650000  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load a table for each assembled transcript with the data\n",
    "exp_df = definitions.get_attribute(prj_assembly_dir / 'ALL.M33262.1.orfanage.tracking.gtf',[\"class_code\",\"ALL_tpm_mean\",\"ALL_num_samples\",\"AY69_tpm_mean\",\"AY69_num_samples\",\"T034_tpm_mean\",\"T034_num_samples\"])\n",
    "# replace - with 0\n",
    "exp_df.replace(\"-\",0,inplace=True)\n",
    "# convert to numeric\n",
    "num_cols = exp_df.columns.difference(['tid','class_code'])\n",
    "exp_df[num_cols] = exp_df[num_cols].astype(float)\n",
    "exp_df.sort_values(by=\"ALL_num_samples\",ascending=False,inplace=True)\n",
    "\n",
    "# compute percent samples \n",
    "exp_df[\"ALL_percent_samples\"] = exp_df[\"ALL_num_samples\"]/metadata_df.shape[0]\n",
    "for monkey in metadata_df[\"animal\"].unique():\n",
    "    exp_df[f\"{monkey}_percent_samples\"] = exp_df[f\"{monkey}_num_samples\"]/metadata_df[metadata_df[\"animal\"]==monkey].shape[0]\n",
    "\n",
    "exp_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "149 novel transcripts detected in the dataset\n",
      "number of novel transcripts with >10% of samples: 4\n"
     ]
    }
   ],
   "source": [
    "# let's get some summary statistics \n",
    "\n",
    "# compute additional information\n",
    "exp_df[\"perc_samples_diff\"] = abs(exp_df[\"AY69_percent_samples\"]-exp_df[\"T034_percent_samples\"])\n",
    "exp_df[\"perc_samples_diff_bin\"] = pd.cut(exp_df[\"perc_samples_diff\"], bins=10, labels=False)\n",
    "\n",
    "# isolate novel\n",
    "novel_exp_df = exp_df[~(exp_df[\"class_code\"]==\"=\")].reset_index(drop=True)\n",
    "# total number of novel transcripts\n",
    "print(f\"{novel_exp_df.shape[0]} novel transcripts detected in the dataset\")\n",
    "\n",
    "# create subsets of most promising novel transcripts\n",
    "tids = novel_exp_df[novel_exp_df[\"ALL_percent_samples\"]>0.1][\"tid\"].tolist()\n",
    "print(f\"number of novel transcripts with >10% of samples: {len(tids)}\")\n",
    "definitions.subset_gtf(prj_assembly_dir / 'ALL.M33262.1.orfanage.tracking.gtf',prj_assembly_dir / 'ALL.M33262.1.orfanage.tracking.top_novel_num_samples.gtf',[],tids)\n",
    "\n",
    "# now repeat by percent difference binned\n",
    "tids = novel_exp_df.sort_values(by=[\"perc_samples_diff_bin\",\"ALL_tpm_mean\"],ascending=False).head(10)[\"tid\"].tolist()\n",
    "definitions.subset_gtf(prj_assembly_dir / 'ALL.M33262.1.orfanage.tracking.gtf',prj_assembly_dir / 'ALL.M33262.1.orfanage.tracking.top_novel_perc_samples_diff.gtf',[],tids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>class_code</th>\n",
       "      <th>ALL_tpm_mean</th>\n",
       "      <th>ALL_num_samples</th>\n",
       "      <th>AY69_tpm_mean</th>\n",
       "      <th>AY69_num_samples</th>\n",
       "      <th>T034_tpm_mean</th>\n",
       "      <th>T034_num_samples</th>\n",
       "      <th>ALL_percent_samples</th>\n",
       "      <th>AY69_percent_samples</th>\n",
       "      <th>T034_percent_samples</th>\n",
       "      <th>perc_samples_diff</th>\n",
       "      <th>perc_samples_diff_bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>ALL_00411274</td>\n",
       "      <td>o</td>\n",
       "      <td>2310.654641</td>\n",
       "      <td>29.0</td>\n",
       "      <td>2310.654641</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.194631</td>\n",
       "      <td>0.325843</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.325843</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>ALL_00411275</td>\n",
       "      <td>j</td>\n",
       "      <td>507.985809</td>\n",
       "      <td>20.0</td>\n",
       "      <td>426.078315</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1245.153259</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.134228</td>\n",
       "      <td>0.202247</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.168914</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>ALL_00411308</td>\n",
       "      <td>j</td>\n",
       "      <td>2555.871491</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2555.871491</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.073826</td>\n",
       "      <td>0.123596</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.123596</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>ALL_00411312</td>\n",
       "      <td>j</td>\n",
       "      <td>2283.966138</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2283.966138</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.033557</td>\n",
       "      <td>0.056180</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.056180</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>ALL_00411289</td>\n",
       "      <td>m</td>\n",
       "      <td>1568.986935</td>\n",
       "      <td>12.0</td>\n",
       "      <td>2385.710388</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1160.625208</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.080537</td>\n",
       "      <td>0.044944</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.088390</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              tid class_code  ALL_tpm_mean  ALL_num_samples  AY69_tpm_mean  \\\n",
       "81   ALL_00411274          o   2310.654641             29.0    2310.654641   \n",
       "51   ALL_00411275          j    507.985809             20.0     426.078315   \n",
       "144  ALL_00411308          j   2555.871491             11.0    2555.871491   \n",
       "141  ALL_00411312          j   2283.966138              5.0    2283.966138   \n",
       "85   ALL_00411289          m   1568.986935             12.0    2385.710388   \n",
       "\n",
       "     AY69_num_samples  T034_tpm_mean  T034_num_samples  ALL_percent_samples  \\\n",
       "81               29.0       0.000000               0.0             0.194631   \n",
       "51               18.0    1245.153259               2.0             0.134228   \n",
       "144              11.0       0.000000               0.0             0.073826   \n",
       "141               5.0       0.000000               0.0             0.033557   \n",
       "85                4.0    1160.625208               8.0             0.080537   \n",
       "\n",
       "     AY69_percent_samples  T034_percent_samples  perc_samples_diff  \\\n",
       "81               0.325843              0.000000           0.325843   \n",
       "51               0.202247              0.033333           0.168914   \n",
       "144              0.123596              0.000000           0.123596   \n",
       "141              0.056180              0.000000           0.056180   \n",
       "85               0.044944              0.133333           0.088390   \n",
       "\n",
       "     perc_samples_diff_bin  \n",
       "81                       7  \n",
       "51                       3  \n",
       "144                      2  \n",
       "141                      1  \n",
       "85                       1  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rank based on the overal tpm and number of samples,\n",
    "# exclusivity to either AY69 or T034\n",
    "# tpm\n",
    "\n",
    "exp_df[\"perc_samples_diff\"] = abs(exp_df[\"AY69_percent_samples\"]-exp_df[\"T034_percent_samples\"])\n",
    "exp_df[\"perc_samples_diff_bin\"] = pd.cut(exp_df[\"perc_samples_diff\"], bins=10, labels=False)\n",
    "exp_df.sort_values(by=[\"perc_samples_diff_bin\",\"ALL_tpm_mean\"],ascending=[False,False],inplace=True)\n",
    "\n",
    "exp_df[~(exp_df[\"class_code\"]==\"=\")].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use stringtie results to load up transcript quantifications \n",
    "strg_tdf = pd.DataFrame()\n",
    "for idx, row in metadata_df.iterrows():\n",
    "    run_id = row['Run']\n",
    "    sample = row['animal']\n",
    "    if not (prj_assembly_dir / f\"{run_id}.gtf\").exists():\n",
    "        continue\n",
    "    \n",
    "    tdf = definitions.get_chains(prj_assembly_dir / f\"{run_id}.gtf\",\"exon\",True)\n",
    "    tmp_df = definitions.get_attribute(prj_assembly_dir / f\"{run_id}.gtf\",[\"TPM\",\"reference_id\"])\n",
    "    tdf = tdf.merge(tmp_df,on=\"tid\")\n",
    "    tdf[\"run_accession\"] = run_id\n",
    "    tdf[\"sample\"] = sample\n",
    "    strg_tdf = pd.concat([strg_tdf, tdf])\n",
    "\n",
    "strg_tdf.to_csv(prj_quant_dir / \"transcript_tpm.tsv\",sep=\"\\t\",index=False)\n",
    "strg_tdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th colspan=\"10\" halign=\"left\">AY69</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">T034</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run_accession</th>\n",
       "      <th>SRR24678268</th>\n",
       "      <th>SRR24678497</th>\n",
       "      <th>SRR24678507</th>\n",
       "      <th>SRR24678531</th>\n",
       "      <th>SRR24678542</th>\n",
       "      <th>SRR24678543</th>\n",
       "      <th>SRR24678550</th>\n",
       "      <th>SRR24678555</th>\n",
       "      <th>SRR24678556</th>\n",
       "      <th>SRR24678567</th>\n",
       "      <th>...</th>\n",
       "      <th>SRR24679215</th>\n",
       "      <th>SRR24679230</th>\n",
       "      <th>SRR24679238</th>\n",
       "      <th>SRR24679260</th>\n",
       "      <th>SRR24679302</th>\n",
       "      <th>SRR24679320</th>\n",
       "      <th>SRR24679338</th>\n",
       "      <th>SRR24679476</th>\n",
       "      <th>SRR24679489</th>\n",
       "      <th>SRR24679514</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reference_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>907.635468</td>\n",
       "      <td>194.610512</td>\n",
       "      <td>1182.871826</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1165.915585</td>\n",
       "      <td>2540.382538</td>\n",
       "      <td>926.502271</td>\n",
       "      <td>540.407471</td>\n",
       "      <td>1496.870127</td>\n",
       "      <td>2116.284363</td>\n",
       "      <td>...</td>\n",
       "      <td>1568.548767</td>\n",
       "      <td>1251.324797</td>\n",
       "      <td>2641.727783</td>\n",
       "      <td>1048.969259</td>\n",
       "      <td>2464.014111</td>\n",
       "      <td>1231.950305</td>\n",
       "      <td>276.303986</td>\n",
       "      <td>1112.275513</td>\n",
       "      <td>803.589813</td>\n",
       "      <td>1107.056024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>env.1</th>\n",
       "      <td>120.444344</td>\n",
       "      <td>31.467482</td>\n",
       "      <td>NaN</td>\n",
       "      <td>109.548798</td>\n",
       "      <td>136.184998</td>\n",
       "      <td>315.649078</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.072433</td>\n",
       "      <td>1129.741821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>891.121277</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>108.138756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>env.2</th>\n",
       "      <td>199.821701</td>\n",
       "      <td>146.941498</td>\n",
       "      <td>46.394848</td>\n",
       "      <td>266.173218</td>\n",
       "      <td>484.943115</td>\n",
       "      <td>512.683228</td>\n",
       "      <td>582.734497</td>\n",
       "      <td>2.795034</td>\n",
       "      <td>1003.000977</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>278.631531</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>391.050659</td>\n",
       "      <td>1132.005493</td>\n",
       "      <td>597.666931</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>223.062332</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>env.3</th>\n",
       "      <td>205.211685</td>\n",
       "      <td>216.750641</td>\n",
       "      <td>47.990814</td>\n",
       "      <td>737.295959</td>\n",
       "      <td>127.695160</td>\n",
       "      <td>426.030273</td>\n",
       "      <td>1469.890991</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1167.700195</td>\n",
       "      <td>508.186615</td>\n",
       "      <td>...</td>\n",
       "      <td>266.988007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>745.718323</td>\n",
       "      <td>1121.547607</td>\n",
       "      <td>607.274109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>163.125336</td>\n",
       "      <td>120.181099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>env.5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>49.857624</td>\n",
       "      <td>101.553154</td>\n",
       "      <td>122.229553</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>347.133789</td>\n",
       "      <td>347.232422</td>\n",
       "      <td>...</td>\n",
       "      <td>266.620636</td>\n",
       "      <td>260.457031</td>\n",
       "      <td>NaN</td>\n",
       "      <td>526.630432</td>\n",
       "      <td>1255.499390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>176.776688</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>env.9</th>\n",
       "      <td>461.366577</td>\n",
       "      <td>188.791214</td>\n",
       "      <td>47.122849</td>\n",
       "      <td>937.056335</td>\n",
       "      <td>74.217941</td>\n",
       "      <td>68.766380</td>\n",
       "      <td>267.621185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1128.697388</td>\n",
       "      <td>47.152004</td>\n",
       "      <td>...</td>\n",
       "      <td>160.718948</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>611.398682</td>\n",
       "      <td>900.808533</td>\n",
       "      <td>416.817322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1422.511353</td>\n",
       "      <td>94.019981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>env/vpu.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>47.298790</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>184.234711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>579.184021</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>363.440216</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>env/vpu.2</th>\n",
       "      <td>155.793671</td>\n",
       "      <td>122.814964</td>\n",
       "      <td>NaN</td>\n",
       "      <td>86.068901</td>\n",
       "      <td>235.735687</td>\n",
       "      <td>606.179871</td>\n",
       "      <td>186.301804</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1043.277344</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>182.682007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>187.249146</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gag.1</th>\n",
       "      <td>92.798599</td>\n",
       "      <td>2744.722168</td>\n",
       "      <td>852.241394</td>\n",
       "      <td>5777.602539</td>\n",
       "      <td>604.631165</td>\n",
       "      <td>397.429993</td>\n",
       "      <td>728.120117</td>\n",
       "      <td>89.112259</td>\n",
       "      <td>2193.041016</td>\n",
       "      <td>225.010208</td>\n",
       "      <td>...</td>\n",
       "      <td>1488.908569</td>\n",
       "      <td>264.517334</td>\n",
       "      <td>167.222153</td>\n",
       "      <td>2949.107910</td>\n",
       "      <td>2268.022949</td>\n",
       "      <td>3706.580322</td>\n",
       "      <td>68.628975</td>\n",
       "      <td>603.426331</td>\n",
       "      <td>279.319244</td>\n",
       "      <td>457.969025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nef.2</th>\n",
       "      <td>176.461441</td>\n",
       "      <td>77.102295</td>\n",
       "      <td>64.800613</td>\n",
       "      <td>495.615173</td>\n",
       "      <td>540.839294</td>\n",
       "      <td>66.814728</td>\n",
       "      <td>404.959351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1067.094727</td>\n",
       "      <td>91.283966</td>\n",
       "      <td>...</td>\n",
       "      <td>376.320557</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>285.560883</td>\n",
       "      <td>1372.876709</td>\n",
       "      <td>2625.879883</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>392.864044</td>\n",
       "      <td>488.903625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nef.3</th>\n",
       "      <td>78.327713</td>\n",
       "      <td>8.525413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43.012287</td>\n",
       "      <td>405.206726</td>\n",
       "      <td>73.878052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>320.997284</td>\n",
       "      <td>391.908081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>311.114594</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>478.976959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nef.4</th>\n",
       "      <td>338.809296</td>\n",
       "      <td>71.799339</td>\n",
       "      <td>37.342960</td>\n",
       "      <td>192.110855</td>\n",
       "      <td>446.958191</td>\n",
       "      <td>268.923035</td>\n",
       "      <td>92.455574</td>\n",
       "      <td>452.487122</td>\n",
       "      <td>775.020386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1195.345459</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.943283</td>\n",
       "      <td>444.252136</td>\n",
       "      <td>442.726013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>659.395813</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nef.5</th>\n",
       "      <td>143.976624</td>\n",
       "      <td>48.009892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34.946964</td>\n",
       "      <td>117.890427</td>\n",
       "      <td>192.385849</td>\n",
       "      <td>22.355080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>117.821045</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>276.073669</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nef.6</th>\n",
       "      <td>197.129501</td>\n",
       "      <td>45.353886</td>\n",
       "      <td>113.749001</td>\n",
       "      <td>491.744324</td>\n",
       "      <td>133.871475</td>\n",
       "      <td>26.775333</td>\n",
       "      <td>120.854729</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2158.757812</td>\n",
       "      <td>7.525825</td>\n",
       "      <td>...</td>\n",
       "      <td>243.508072</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>169.380569</td>\n",
       "      <td>550.601440</td>\n",
       "      <td>371.789459</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2877.757568</td>\n",
       "      <td>132.476395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pol.1</th>\n",
       "      <td>21.448082</td>\n",
       "      <td>885.406616</td>\n",
       "      <td>10291.432617</td>\n",
       "      <td>2082.374023</td>\n",
       "      <td>418.289185</td>\n",
       "      <td>272.767639</td>\n",
       "      <td>575.501831</td>\n",
       "      <td>256.261414</td>\n",
       "      <td>2304.765381</td>\n",
       "      <td>214.430801</td>\n",
       "      <td>...</td>\n",
       "      <td>1078.468628</td>\n",
       "      <td>279.440765</td>\n",
       "      <td>162.884949</td>\n",
       "      <td>2006.891724</td>\n",
       "      <td>3547.266846</td>\n",
       "      <td>2103.467041</td>\n",
       "      <td>1139.304688</td>\n",
       "      <td>822.342529</td>\n",
       "      <td>364.535034</td>\n",
       "      <td>485.568878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rev.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>55.566547</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>186.270599</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>583.354309</td>\n",
       "      <td>185.145462</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1266.786133</td>\n",
       "      <td>826.344482</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>403.601654</td>\n",
       "      <td>448.632294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rev.7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>52.811993</td>\n",
       "      <td>29.443998</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>401.890167</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>536.749268</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>299.949341</td>\n",
       "      <td>587.899048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>506.400818</td>\n",
       "      <td>542.790527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rev.8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.373375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>56.105366</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>121.352928</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.160088</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rev.9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>26.326689</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43.731178</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>526.108459</td>\n",
       "      <td>138.673004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>166.959595</td>\n",
       "      <td>118.040947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1586.297485</td>\n",
       "      <td>80.596626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tat.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28.494141</td>\n",
       "      <td>51.723976</td>\n",
       "      <td>129.134125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>179.347763</td>\n",
       "      <td>57.465328</td>\n",
       "      <td>...</td>\n",
       "      <td>515.724670</td>\n",
       "      <td>170.121536</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72.819054</td>\n",
       "      <td>312.389526</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>381.894714</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tat.2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>106.038040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>213.905365</td>\n",
       "      <td>39.742493</td>\n",
       "      <td>...</td>\n",
       "      <td>2239.987061</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>497.011017</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tat.3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>45.554169</td>\n",
       "      <td>57.927528</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>325.303680</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>336.738739</td>\n",
       "      <td>47.864845</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>233.956116</td>\n",
       "      <td>NaN</td>\n",
       "      <td>395.772858</td>\n",
       "      <td>891.436523</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>348.678162</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tat.4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>12.428283</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>191.720428</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61.986370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>45.252041</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.350533</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tat.7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21.534731</td>\n",
       "      <td>37.918331</td>\n",
       "      <td>61.892872</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.637054</td>\n",
       "      <td>6.416903</td>\n",
       "      <td>...</td>\n",
       "      <td>582.050110</td>\n",
       "      <td>129.505203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.294735</td>\n",
       "      <td>181.553421</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>193.391312</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vif.2</th>\n",
       "      <td>2.949473</td>\n",
       "      <td>289.391876</td>\n",
       "      <td>72.694153</td>\n",
       "      <td>114.496872</td>\n",
       "      <td>77.181419</td>\n",
       "      <td>402.387085</td>\n",
       "      <td>938.189392</td>\n",
       "      <td>1.112235</td>\n",
       "      <td>328.502991</td>\n",
       "      <td>175.778870</td>\n",
       "      <td>...</td>\n",
       "      <td>226.402512</td>\n",
       "      <td>128.093185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36.883396</td>\n",
       "      <td>827.850708</td>\n",
       "      <td>623.365479</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.881691</td>\n",
       "      <td>1284.017578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vif.3</th>\n",
       "      <td>1.548109</td>\n",
       "      <td>224.684631</td>\n",
       "      <td>42.061466</td>\n",
       "      <td>100.769241</td>\n",
       "      <td>37.250889</td>\n",
       "      <td>118.685089</td>\n",
       "      <td>174.882797</td>\n",
       "      <td>25.155865</td>\n",
       "      <td>497.482697</td>\n",
       "      <td>9.548364</td>\n",
       "      <td>...</td>\n",
       "      <td>949.244751</td>\n",
       "      <td>98.189926</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23.837240</td>\n",
       "      <td>801.350464</td>\n",
       "      <td>485.702820</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.728075</td>\n",
       "      <td>85.667290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vpr.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35.608200</td>\n",
       "      <td>56.104286</td>\n",
       "      <td>153.000107</td>\n",
       "      <td>138.672089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>179.943314</td>\n",
       "      <td>41.761539</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61.032024</td>\n",
       "      <td>181.690567</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>252.249313</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vpr.2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30.360949</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>156.856171</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>218.773788</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vpr.3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57.684296</td>\n",
       "      <td>92.211906</td>\n",
       "      <td>108.144249</td>\n",
       "      <td>410.954315</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>195.623138</td>\n",
       "      <td>203.167984</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>305.943298</td>\n",
       "      <td>756.458008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2814.114990</td>\n",
       "      <td>179.481293</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vpr.4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46.676311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>739.657410</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>188.285965</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vpx.1</th>\n",
       "      <td>13.767508</td>\n",
       "      <td>121.789665</td>\n",
       "      <td>88.054878</td>\n",
       "      <td>380.265686</td>\n",
       "      <td>110.206413</td>\n",
       "      <td>524.412720</td>\n",
       "      <td>203.330887</td>\n",
       "      <td>0.653416</td>\n",
       "      <td>859.525574</td>\n",
       "      <td>324.841553</td>\n",
       "      <td>...</td>\n",
       "      <td>252.166122</td>\n",
       "      <td>146.760788</td>\n",
       "      <td>NaN</td>\n",
       "      <td>295.850861</td>\n",
       "      <td>1098.337524</td>\n",
       "      <td>843.304260</td>\n",
       "      <td>11.492917</td>\n",
       "      <td>174.281403</td>\n",
       "      <td>189.305222</td>\n",
       "      <td>112.668465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vpx.2</th>\n",
       "      <td>4.695217</td>\n",
       "      <td>25.850071</td>\n",
       "      <td>NaN</td>\n",
       "      <td>66.508400</td>\n",
       "      <td>120.091698</td>\n",
       "      <td>410.990997</td>\n",
       "      <td>55.025246</td>\n",
       "      <td>NaN</td>\n",
       "      <td>175.227142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>469.445068</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>193.740860</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vpx.3</th>\n",
       "      <td>7.194247</td>\n",
       "      <td>96.322395</td>\n",
       "      <td>67.586304</td>\n",
       "      <td>417.288239</td>\n",
       "      <td>30.946638</td>\n",
       "      <td>96.138496</td>\n",
       "      <td>37.189949</td>\n",
       "      <td>1.045738</td>\n",
       "      <td>1115.136230</td>\n",
       "      <td>19.167196</td>\n",
       "      <td>...</td>\n",
       "      <td>166.089233</td>\n",
       "      <td>120.211601</td>\n",
       "      <td>NaN</td>\n",
       "      <td>114.935127</td>\n",
       "      <td>1043.695557</td>\n",
       "      <td>552.853577</td>\n",
       "      <td>NaN</td>\n",
       "      <td>78.343132</td>\n",
       "      <td>320.587708</td>\n",
       "      <td>93.140938</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33 rows  149 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample               AY69                                          \\\n",
       "run_accession SRR24678268  SRR24678497   SRR24678507  SRR24678531   \n",
       "reference_id                                                        \n",
       "-              907.635468   194.610512   1182.871826          NaN   \n",
       "env.1          120.444344    31.467482           NaN   109.548798   \n",
       "env.2          199.821701   146.941498     46.394848   266.173218   \n",
       "env.3          205.211685   216.750641     47.990814   737.295959   \n",
       "env.5                 NaN          NaN     49.857624   101.553154   \n",
       "env.9          461.366577   188.791214     47.122849   937.056335   \n",
       "env/vpu.1             NaN          NaN           NaN    47.298790   \n",
       "env/vpu.2      155.793671   122.814964           NaN    86.068901   \n",
       "gag.1           92.798599  2744.722168    852.241394  5777.602539   \n",
       "nef.2          176.461441    77.102295     64.800613   495.615173   \n",
       "nef.3           78.327713     8.525413           NaN    43.012287   \n",
       "nef.4          338.809296    71.799339     37.342960   192.110855   \n",
       "nef.5          143.976624    48.009892           NaN    34.946964   \n",
       "nef.6          197.129501    45.353886    113.749001   491.744324   \n",
       "pol.1           21.448082   885.406616  10291.432617  2082.374023   \n",
       "rev.1                 NaN    55.566547           NaN          NaN   \n",
       "rev.7                 NaN          NaN     52.811993    29.443998   \n",
       "rev.8                 NaN          NaN           NaN    25.373375   \n",
       "rev.9                 NaN    26.326689           NaN          NaN   \n",
       "tat.1                 NaN          NaN     28.494141    51.723976   \n",
       "tat.2                 NaN          NaN           NaN          NaN   \n",
       "tat.3                 NaN    45.554169     57.927528          NaN   \n",
       "tat.4                 NaN    12.428283           NaN          NaN   \n",
       "tat.7                 NaN          NaN     21.534731    37.918331   \n",
       "vif.2            2.949473   289.391876     72.694153   114.496872   \n",
       "vif.3            1.548109   224.684631     42.061466   100.769241   \n",
       "vpr.1                 NaN          NaN     35.608200    56.104286   \n",
       "vpr.2                 NaN          NaN           NaN    30.360949   \n",
       "vpr.3                 NaN          NaN     57.684296    92.211906   \n",
       "vpr.4                 NaN          NaN           NaN    46.676311   \n",
       "vpx.1           13.767508   121.789665     88.054878   380.265686   \n",
       "vpx.2            4.695217    25.850071           NaN    66.508400   \n",
       "vpx.3            7.194247    96.322395     67.586304   417.288239   \n",
       "\n",
       "sample                                                                         \\\n",
       "run_accession  SRR24678542  SRR24678543  SRR24678550 SRR24678555  SRR24678556   \n",
       "reference_id                                                                    \n",
       "-              1165.915585  2540.382538   926.502271  540.407471  1496.870127   \n",
       "env.1           136.184998   315.649078          NaN    2.072433  1129.741821   \n",
       "env.2           484.943115   512.683228   582.734497    2.795034  1003.000977   \n",
       "env.3           127.695160   426.030273  1469.890991         NaN  1167.700195   \n",
       "env.5           122.229553          NaN          NaN         NaN   347.133789   \n",
       "env.9            74.217941    68.766380   267.621185         NaN  1128.697388   \n",
       "env/vpu.1              NaN          NaN          NaN         NaN   184.234711   \n",
       "env/vpu.2       235.735687   606.179871   186.301804         NaN  1043.277344   \n",
       "gag.1           604.631165   397.429993   728.120117   89.112259  2193.041016   \n",
       "nef.2           540.839294    66.814728   404.959351         NaN  1067.094727   \n",
       "nef.3           405.206726    73.878052          NaN  320.997284   391.908081   \n",
       "nef.4           446.958191   268.923035    92.455574  452.487122   775.020386   \n",
       "nef.5           117.890427   192.385849    22.355080         NaN   117.821045   \n",
       "nef.6           133.871475    26.775333   120.854729         NaN  2158.757812   \n",
       "pol.1           418.289185   272.767639   575.501831  256.261414  2304.765381   \n",
       "rev.1                  NaN          NaN          NaN         NaN   186.270599   \n",
       "rev.7                  NaN          NaN          NaN         NaN   401.890167   \n",
       "rev.8                  NaN          NaN          NaN         NaN    56.105366   \n",
       "rev.9                  NaN          NaN          NaN         NaN    43.731178   \n",
       "tat.1           129.134125          NaN          NaN         NaN   179.347763   \n",
       "tat.2                  NaN          NaN   106.038040         NaN   213.905365   \n",
       "tat.3                  NaN   325.303680          NaN         NaN   336.738739   \n",
       "tat.4                  NaN   191.720428          NaN         NaN    61.986370   \n",
       "tat.7            61.892872          NaN          NaN         NaN   100.637054   \n",
       "vif.2            77.181419   402.387085   938.189392    1.112235   328.502991   \n",
       "vif.3            37.250889   118.685089   174.882797   25.155865   497.482697   \n",
       "vpr.1           153.000107   138.672089          NaN         NaN   179.943314   \n",
       "vpr.2                  NaN          NaN          NaN         NaN          NaN   \n",
       "vpr.3           108.144249   410.954315          NaN         NaN   195.623138   \n",
       "vpr.4                  NaN          NaN          NaN         NaN          NaN   \n",
       "vpx.1           110.206413   524.412720   203.330887    0.653416   859.525574   \n",
       "vpx.2           120.091698   410.990997    55.025246         NaN   175.227142   \n",
       "vpx.3            30.946638    96.138496    37.189949    1.045738  1115.136230   \n",
       "\n",
       "sample                      ...         T034                            \\\n",
       "run_accession  SRR24678567  ...  SRR24679215  SRR24679230  SRR24679238   \n",
       "reference_id                ...                                          \n",
       "-              2116.284363  ...  1568.548767  1251.324797  2641.727783   \n",
       "env.1                  NaN  ...          NaN          NaN          NaN   \n",
       "env.2                  NaN  ...   278.631531          NaN          NaN   \n",
       "env.3           508.186615  ...   266.988007          NaN          NaN   \n",
       "env.5           347.232422  ...   266.620636   260.457031          NaN   \n",
       "env.9            47.152004  ...   160.718948          NaN          NaN   \n",
       "env/vpu.1              NaN  ...          NaN          NaN          NaN   \n",
       "env/vpu.2              NaN  ...          NaN          NaN          NaN   \n",
       "gag.1           225.010208  ...  1488.908569   264.517334   167.222153   \n",
       "nef.2            91.283966  ...   376.320557          NaN          NaN   \n",
       "nef.3                  NaN  ...          NaN          NaN          NaN   \n",
       "nef.4                  NaN  ...  1195.345459          NaN          NaN   \n",
       "nef.5                  NaN  ...          NaN          NaN          NaN   \n",
       "nef.6             7.525825  ...   243.508072          NaN          NaN   \n",
       "pol.1           214.430801  ...  1078.468628   279.440765   162.884949   \n",
       "rev.1                  NaN  ...   583.354309   185.145462          NaN   \n",
       "rev.7                  NaN  ...   536.749268          NaN          NaN   \n",
       "rev.8                  NaN  ...          NaN          NaN          NaN   \n",
       "rev.9                  NaN  ...   526.108459   138.673004          NaN   \n",
       "tat.1            57.465328  ...   515.724670   170.121536          NaN   \n",
       "tat.2            39.742493  ...  2239.987061          NaN          NaN   \n",
       "tat.3            47.864845  ...          NaN   233.956116          NaN   \n",
       "tat.4                  NaN  ...          NaN          NaN          NaN   \n",
       "tat.7             6.416903  ...   582.050110   129.505203          NaN   \n",
       "vif.2           175.778870  ...   226.402512   128.093185          NaN   \n",
       "vif.3             9.548364  ...   949.244751    98.189926          NaN   \n",
       "vpr.1            41.761539  ...          NaN          NaN          NaN   \n",
       "vpr.2                  NaN  ...          NaN          NaN          NaN   \n",
       "vpr.3           203.167984  ...          NaN          NaN          NaN   \n",
       "vpr.4                  NaN  ...          NaN          NaN          NaN   \n",
       "vpx.1           324.841553  ...   252.166122   146.760788          NaN   \n",
       "vpx.2                  NaN  ...          NaN          NaN          NaN   \n",
       "vpx.3            19.167196  ...   166.089233   120.211601          NaN   \n",
       "\n",
       "sample                                                             \\\n",
       "run_accession  SRR24679260  SRR24679302  SRR24679320  SRR24679338   \n",
       "reference_id                                                        \n",
       "-              1048.969259  2464.014111  1231.950305   276.303986   \n",
       "env.1                  NaN   891.121277          NaN          NaN   \n",
       "env.2           391.050659  1132.005493   597.666931          NaN   \n",
       "env.3           745.718323  1121.547607   607.274109          NaN   \n",
       "env.5           526.630432  1255.499390          NaN          NaN   \n",
       "env.9           611.398682   900.808533   416.817322          NaN   \n",
       "env/vpu.1       579.184021          NaN          NaN          NaN   \n",
       "env/vpu.2       182.682007          NaN          NaN          NaN   \n",
       "gag.1          2949.107910  2268.022949  3706.580322    68.628975   \n",
       "nef.2           285.560883  1372.876709  2625.879883          NaN   \n",
       "nef.3                  NaN   311.114594          NaN          NaN   \n",
       "nef.4            41.943283   444.252136   442.726013          NaN   \n",
       "nef.5                  NaN          NaN          NaN          NaN   \n",
       "nef.6           169.380569   550.601440   371.789459          NaN   \n",
       "pol.1          2006.891724  3547.266846  2103.467041  1139.304688   \n",
       "rev.1                  NaN  1266.786133   826.344482          NaN   \n",
       "rev.7           299.949341   587.899048          NaN          NaN   \n",
       "rev.8           121.352928          NaN          NaN          NaN   \n",
       "rev.9                  NaN   166.959595   118.040947          NaN   \n",
       "tat.1            72.819054   312.389526          NaN          NaN   \n",
       "tat.2                  NaN   497.011017          NaN          NaN   \n",
       "tat.3           395.772858   891.436523          NaN          NaN   \n",
       "tat.4            45.252041          NaN          NaN          NaN   \n",
       "tat.7            41.294735   181.553421          NaN          NaN   \n",
       "vif.2            36.883396   827.850708   623.365479          NaN   \n",
       "vif.3            23.837240   801.350464   485.702820          NaN   \n",
       "vpr.1            61.032024   181.690567          NaN          NaN   \n",
       "vpr.2                  NaN   156.856171          NaN          NaN   \n",
       "vpr.3           305.943298   756.458008          NaN          NaN   \n",
       "vpr.4                  NaN   739.657410          NaN          NaN   \n",
       "vpx.1           295.850861  1098.337524   843.304260    11.492917   \n",
       "vpx.2           469.445068          NaN          NaN          NaN   \n",
       "vpx.3           114.935127  1043.695557   552.853577          NaN   \n",
       "\n",
       "sample                                                \n",
       "run_accession  SRR24679476  SRR24679489  SRR24679514  \n",
       "reference_id                                          \n",
       "-              1112.275513   803.589813  1107.056024  \n",
       "env.1                  NaN          NaN   108.138756  \n",
       "env.2                  NaN   223.062332          NaN  \n",
       "env.3                  NaN   163.125336   120.181099  \n",
       "env.5                  NaN   176.776688          NaN  \n",
       "env.9                  NaN  1422.511353    94.019981  \n",
       "env/vpu.1              NaN   363.440216          NaN  \n",
       "env/vpu.2              NaN   187.249146          NaN  \n",
       "gag.1           603.426331   279.319244   457.969025  \n",
       "nef.2                  NaN   392.864044   488.903625  \n",
       "nef.3                  NaN          NaN   478.976959  \n",
       "nef.4                  NaN   659.395813          NaN  \n",
       "nef.5                  NaN   276.073669          NaN  \n",
       "nef.6                  NaN  2877.757568   132.476395  \n",
       "pol.1           822.342529   364.535034   485.568878  \n",
       "rev.1                  NaN   403.601654   448.632294  \n",
       "rev.7                  NaN   506.400818   542.790527  \n",
       "rev.8                  NaN    99.160088          NaN  \n",
       "rev.9                  NaN  1586.297485    80.596626  \n",
       "tat.1                  NaN   381.894714          NaN  \n",
       "tat.2                  NaN          NaN          NaN  \n",
       "tat.3                  NaN   348.678162          NaN  \n",
       "tat.4                  NaN    99.350533          NaN  \n",
       "tat.7                  NaN   193.391312          NaN  \n",
       "vif.2                  NaN    74.881691  1284.017578  \n",
       "vif.3                  NaN    18.728075    85.667290  \n",
       "vpr.1                  NaN   252.249313          NaN  \n",
       "vpr.2                  NaN   218.773788          NaN  \n",
       "vpr.3          2814.114990   179.481293          NaN  \n",
       "vpr.4                  NaN   188.285965          NaN  \n",
       "vpx.1           174.281403   189.305222   112.668465  \n",
       "vpx.2                  NaN   193.740860          NaN  \n",
       "vpx.3            78.343132   320.587708    93.140938  \n",
       "\n",
       "[33 rows x 149 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strg_tdf = pd.read_csv(prj_quant_dir / \"transcript_tpm.tsv\", sep='\\t')\n",
    "\n",
    "strg_tdf = strg_tdf[(strg_tdf[\"seqid\"]==\"M33262.1\")].reset_index(drop=True)\n",
    "\n",
    "# Pivot table for statistical analysis\n",
    "pivot_df = strg_tdf.pivot_table(index='reference_id', columns=['sample',\"run_accession\"], values='TPM')\n",
    "pivot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-</td>\n",
       "      <td>0.744492</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>env.1</td>\n",
       "      <td>0.477565</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>env.2</td>\n",
       "      <td>0.518538</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>env.3</td>\n",
       "      <td>0.677672</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>env.5</td>\n",
       "      <td>0.062076</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>env.9</td>\n",
       "      <td>0.181223</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>env/vpu.1</td>\n",
       "      <td>0.017150</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>env/vpu.2</td>\n",
       "      <td>0.913453</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>gag.1</td>\n",
       "      <td>0.037110</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>nef.2</td>\n",
       "      <td>0.137620</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>nef.3</td>\n",
       "      <td>0.077421</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>nef.4</td>\n",
       "      <td>0.099827</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>nef.5</td>\n",
       "      <td>0.201293</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>nef.6</td>\n",
       "      <td>0.209021</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>pol.1</td>\n",
       "      <td>0.174419</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rev.1</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>rev.7</td>\n",
       "      <td>0.001894</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>rev.8</td>\n",
       "      <td>0.037022</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>rev.9</td>\n",
       "      <td>0.050635</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>tat.1</td>\n",
       "      <td>0.002995</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>tat.2</td>\n",
       "      <td>0.000803</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>tat.3</td>\n",
       "      <td>0.067134</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>tat.4</td>\n",
       "      <td>0.936128</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>tat.7</td>\n",
       "      <td>0.000460</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>vif.2</td>\n",
       "      <td>0.002221</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>vif.3</td>\n",
       "      <td>0.248988</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>vpr.1</td>\n",
       "      <td>0.315243</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>vpr.2</td>\n",
       "      <td>0.000191</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>vpr.3</td>\n",
       "      <td>0.095033</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>vpr.4</td>\n",
       "      <td>0.064408</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>vpx.1</td>\n",
       "      <td>0.813003</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>vpx.2</td>\n",
       "      <td>0.022182</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>vpx.3</td>\n",
       "      <td>0.378664</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   reference_id      pval  significant\n",
       "0             -  0.744492        False\n",
       "1         env.1  0.477565        False\n",
       "2         env.2  0.518538        False\n",
       "3         env.3  0.677672        False\n",
       "4         env.5  0.062076        False\n",
       "5         env.9  0.181223        False\n",
       "6     env/vpu.1  0.017150         True\n",
       "7     env/vpu.2  0.913453        False\n",
       "8         gag.1  0.037110         True\n",
       "9         nef.2  0.137620        False\n",
       "10        nef.3  0.077421        False\n",
       "11        nef.4  0.099827        False\n",
       "12        nef.5  0.201293        False\n",
       "13        nef.6  0.209021        False\n",
       "14        pol.1  0.174419        False\n",
       "15        rev.1  0.000346         True\n",
       "16        rev.7  0.001894         True\n",
       "17        rev.8  0.037022         True\n",
       "18        rev.9  0.050635        False\n",
       "19        tat.1  0.002995         True\n",
       "20        tat.2  0.000803         True\n",
       "21        tat.3  0.067134        False\n",
       "22        tat.4  0.936128        False\n",
       "23        tat.7  0.000460         True\n",
       "24        vif.2  0.002221         True\n",
       "25        vif.3  0.248988        False\n",
       "26        vpr.1  0.315243        False\n",
       "27        vpr.2  0.000191         True\n",
       "28        vpr.3  0.095033        False\n",
       "29        vpr.4  0.064408        False\n",
       "30        vpx.1  0.813003        False\n",
       "31        vpx.2  0.022182         True\n",
       "32        vpx.3  0.378664        False"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Perform ANOVA to identify significant transcripts\n",
    "anova_results = []\n",
    "for tid, row in pivot_df.iterrows():\n",
    "    groups = [\n",
    "        row.filter(like=\"AY69\").dropna(),\n",
    "        row.filter(like=\"T034\").dropna()\n",
    "    ]\n",
    "    if all(len(group) > 1 for group in groups):  # Ensure replicates are present\n",
    "        stat, pval = f_oneway(*groups)\n",
    "        anova_results.append((tid, pval))\n",
    "\n",
    "anova_df = pd.DataFrame(anova_results, columns=['reference_id', 'pval'])\n",
    "anova_df['significant'] = anova_df['pval'] < 0.05\n",
    "anova_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_significant_tids = significant_tids[:5]\n",
    "\n",
    "# Filter data for these transcripts\n",
    "top_significant_data = strg_tdf[strg_tdf['reference_id'].isin(top_significant_tids)]\n",
    "\n",
    "# Create a new column for groups based on \"sample\"\n",
    "top_significant_data['group'] = top_significant_data['sample'].apply(lambda x: 'AY69' if 'AY69' in x else 'T034')\n",
    "\n",
    "# Number of transcripts to plot\n",
    "num_transcripts = 5  # Adjust as needed\n",
    "top_transcripts = significant_tids[:num_transcripts]\n",
    "\n",
    "# Create a figure with subplots\n",
    "fig, axes = plt.subplots(1, num_transcripts, figsize=(4*num_transcripts, 6), sharey=False)\n",
    "\n",
    "# Plot each transcript as a separate violin plot\n",
    "for i, tid in enumerate(top_transcripts):\n",
    "    transcript_data = strg_tdf[strg_tdf['reference_id'] == tid]\n",
    "    \n",
    "    # Plot on the appropriate subplot\n",
    "    ax = axes[i] if num_transcripts > 1 else axes\n",
    "    \n",
    "    # Create the split violin plot\n",
    "    sns.violinplot(\n",
    "        data=transcript_data,\n",
    "        x='reference_id',  # This will be constant within each subplot\n",
    "        y='TPM',\n",
    "        hue='sample',\n",
    "        split=True,\n",
    "        inner='quart',\n",
    "        fill=False,\n",
    "        ax=ax\n",
    "    )\n",
    "    \n",
    "    # Customize the subplot\n",
    "    ax.set_title(f\"{tid}\")\n",
    "    ax.set_xlabel(\"\")  # Remove x label as it's redundant\n",
    "    \n",
    "    # Only keep y-label for the first plot\n",
    "    if i > 0:\n",
    "        ax.set_ylabel(\"\")\n",
    "    \n",
    "    # Remove legend from all but the last plot\n",
    "    if i < num_transcripts - 1:\n",
    "        ax.get_legend().remove()\n",
    "\n",
    "# Adjust layout\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write a list of tiecov coverage and junction files and names of monkeys for the cov/sj/tn params of sashimi\n",
    "cov_lst_fname = prj_aux_dir / 'cov_lst.txt'\n",
    "with open(cov_lst_fname, 'w') as outFP:\n",
    "    for monkey in metadata_df[\"animal\"].unique():\n",
    "        outFP.write(f\"{prj_aln_dir}/{monkey}.tb.M33262.1.coverage.bedgraph\\n\")\n",
    "sj_lst_fname = prj_aux_dir / 'sj_lst.txt'\n",
    "with open(sj_lst_fname, 'w') as outFP:\n",
    "    for monkey in metadata_df[\"animal\"].unique():\n",
    "        outFP.write(f\"{prj_aln_dir}/{monkey}.tb.M33262.1.junctions.bed\\n\")\n",
    "tn_lst_fname = prj_aux_dir / 'tn_lst.txt'\n",
    "with open(tn_lst_fname, 'w') as outFP:\n",
    "    for monkey in metadata_df[\"animal\"].unique():\n",
    "        outFP.write(f\"{monkey}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build sashimi plots for the significant transcripts\n",
    "# we want to compare tiebrush between T034 and Ay69 for a given transcript only\n",
    "\n",
    "for tid in set(strg_tdf['reference_id'].tolist()):  # Show top 5 significant transcripts\n",
    "    try:\n",
    "        tx_gtf_fname = prj_aux_dir / f'{tid}.gtf'\n",
    "        tx_svg_fname = prj_aux_dir / f'{tid}.svg'\n",
    "\n",
    "        # extract the transcript from the gtf\n",
    "        definitions.subset_gtf(reference_gtf_fname,tx_gtf_fname,False,[tid])\n",
    "\n",
    "        # build sashimi plot\n",
    "        sashimi_cmd = [sashimi_bin,\n",
    "                        \"--title\",tid,\n",
    "                        \"--gtf\",str(tx_gtf_fname),\n",
    "                        \"-o\",str(tx_svg_fname),\n",
    "                        \"--normalize\",\n",
    "                        \"--subtract\",\"0\",\n",
    "                        \"--intron_scale\",\"1\",\n",
    "                        \"--exon_scale\",\"1\",\n",
    "                        \"--tn\",str(tn_lst_fname),\n",
    "                        \"--cov\",str(cov_lst_fname),\n",
    "                        \"--sj\",str(sj_lst_fname)]\n",
    "        print(\" \".join(sashimi_cmd))\n",
    "        subprocess.call(sashimi_cmd)\n",
    "\n",
    "        # display the svg within notebook\n",
    "        display(SVG(filename=tx_svg_fname))\n",
    "\n",
    "\n",
    "        # SASHIMI WITHOUT NORMALIZATION\n",
    "        tx_svg_fname = prj_aux_dir / f'{tid}.non_normalized.svg'\n",
    "\n",
    "        sashimi_cmd = [sashimi_bin,\n",
    "                        \"--title\",tid,\n",
    "                        \"--gtf\",str(tx_gtf_fname),\n",
    "                        \"-o\",str(tx_svg_fname),\n",
    "                        \"--intron_scale\",\"1\",\n",
    "                        \"--exon_scale\",\"1\",\n",
    "                        \"--tn\",str(tn_lst_fname),\n",
    "                        \"--cov\",str(cov_lst_fname),\n",
    "                        \"--sj\",str(sj_lst_fname)]\n",
    "        print(\" \".join(sashimi_cmd))\n",
    "        subprocess.call(sashimi_cmd)\n",
    "\n",
    "        # display the svg within notebook\n",
    "        display(SVG(filename=tx_svg_fname))\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing {tid}: {e}\")\n",
    "        continue\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load transcriptomes for each monkey and investigate novel transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# examine tiebrush outputs for each monkey"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hivAtlas",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
