{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "project page: https://www.ncbi.nlm.nih.gov/bioproject/PRJNA287649\n",
    "study: https://retrovirology.biomedcentral.com/articles/10.1186/s12977-015-0204-2\n",
    "biosample_results: https://www.ncbi.nlm.nih.gov/biosample?Db=biosample&DbFrom=bioproject&Cmd=Link&LinkName=bioproject_biosample&LinkReadableName=BioSample&ordinalpos=1&IdsFromResult=287649"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/avaraby1/miniconda3/envs/hivAtlas/lib/python3.11/site-packages/pysradb/utils.py:14: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from tqdm.autonotebook import tqdm\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import time\n",
    "import random\n",
    "import requests\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "from typing import List, Union\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from pysradb import SRAweb\n",
    "\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = Path.cwd().parent.parent.parent / 'HIV_Atlas_Creation' / 'data'\n",
    "\n",
    "sequence_dir = data_dir / 'sequences'\n",
    "assert sequence_dir.exists(), f\"sequence_dir does not exist: {sequence_dir}\"\n",
    "\n",
    "annotation_dir = data_dir / 'annotation'\n",
    "assert annotation_dir.exists(), f\"annotation_dir does not exist: {annotation_dir}\"\n",
    "\n",
    "reference_fasta_fname = sequence_dir / 'K03454.fasta'\n",
    "reference_gtf_fname = annotation_dir / 'K03454/K03454.vira.gtf'\n",
    "\n",
    "base_dir = Path.cwd().parent.parent\n",
    "\n",
    "soft_dir = base_dir / 'soft'\n",
    "\n",
    "prj_id = 'PRJNA287649'\n",
    "\n",
    "outdir = base_dir / 'results' / prj_id\n",
    "outdir.mkdir(parents=True, exist_ok=True)\n",
    "prj_data_dir = outdir / 'data'\n",
    "prj_data_dir.mkdir(parents=True, exist_ok=True)\n",
    "prj_fastq_dir = prj_data_dir / 'fastq'\n",
    "prj_fastq_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "prj_aln_dir = outdir / 'alignment'\n",
    "prj_aln_dir.mkdir(parents=True, exist_ok=True)\n",
    "prj_assembly_dir = outdir / 'assembly'\n",
    "prj_assembly_dir.mkdir(parents=True, exist_ok=True)\n",
    "prj_quant_dir = outdir / 'quantification'\n",
    "prj_quant_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "rsem_idx_basename = base_dir / 'data' / 'hg38_K03454_rsem'\n",
    "mudskipper_idx_basename = base_dir / 'data' / 'hg38_K03454.mudskipper'\n",
    "hisat_idx_basename = base_dir / 'data' / 'hg38_K03454'\n",
    "reference_fasta_fname = base_dir / 'data' / 'hg38_K03454.fasta'\n",
    "reference_gtf_fname = base_dir / 'data' / 'hg38_K03454.gtf'\n",
    "t2g_fname = base_dir / 'data' / 'hg38_K03454.t2g.tsv'\n",
    "\n",
    "biosample_results_fname = prj_data_dir / 'metadata/biosample_results.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 1\n",
    "\n",
    "sys.path.insert(0, str(soft_dir / \"genomic_scripts\"))\n",
    "%aimport definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>biosample</th>\n",
       "      <th>sra</th>\n",
       "      <th>geo</th>\n",
       "      <th>run_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Donor 36 CD14+ mono</td>\n",
       "      <td>SAMN03785028</td>\n",
       "      <td>SRS967589</td>\n",
       "      <td>GSM1717162</td>\n",
       "      <td>SRR2072610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Donor 36 SLAN DC</td>\n",
       "      <td>SAMN03785027</td>\n",
       "      <td>SRS967590</td>\n",
       "      <td>GSM1717161</td>\n",
       "      <td>SRR2072609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Donor 36 CD1c+ mDC</td>\n",
       "      <td>SAMN03785026</td>\n",
       "      <td>SRS967591</td>\n",
       "      <td>GSM1717160</td>\n",
       "      <td>SRR2072608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Donor 36 pDC</td>\n",
       "      <td>SAMN03785025</td>\n",
       "      <td>SRS967593</td>\n",
       "      <td>GSM1717159</td>\n",
       "      <td>SRR2072607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Donor 28 CD14+ Mono</td>\n",
       "      <td>SAMN03785024</td>\n",
       "      <td>SRS967592</td>\n",
       "      <td>GSM1717158</td>\n",
       "      <td>SRR2072606</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                sample     biosample        sra         geo run_accession\n",
       "0  Donor 36 CD14+ mono  SAMN03785028  SRS967589  GSM1717162    SRR2072610\n",
       "1     Donor 36 SLAN DC  SAMN03785027  SRS967590  GSM1717161    SRR2072609\n",
       "2   Donor 36 CD1c+ mDC  SAMN03785026  SRS967591  GSM1717160    SRR2072608\n",
       "3         Donor 36 pDC  SAMN03785025  SRS967593  GSM1717159    SRR2072607\n",
       "4  Donor 28 CD14+ Mono  SAMN03785024  SRS967592  GSM1717158    SRR2072606"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load metadata\n",
    "metadata_df = pd.read_csv(biosample_results_fname, sep='\\t')\n",
    "\n",
    "db = SRAweb()\n",
    "batch_results = db.sra_metadata(metadata_df[\"sra\"].tolist(), detailed=True)\n",
    "metadata_df = metadata_df.merge(batch_results[[\"run_accession\",\"sample_accession\"]],left_on=\"sra\",right_on=\"sample_accession\",how=\"left\")\n",
    "metadata_df = metadata_df.drop(columns=[\"sample_accession\"])\n",
    "metadata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract column with experimental information\n",
    "metadata_df[\"cell_cat\"] = metadata_df[\"sample\"].str.split(\" \",n=2,expand=True)[2]\n",
    "metadata_df[\"donor_cat\"] = metadata_df[\"sample\"].str.split(' ').str[:2].str.join(' ')\n",
    "metadata_df.to_csv(prj_data_dir / 'metadata/metadata.tsv',sep=\"\\t\",index=False)\n",
    "\n",
    "pheno_df = metadata_df[[\"run_accession\",\"donor_cat\",\"cell_cat\"]]\n",
    "pheno_df.columns = [\"id\",\"donor_cat\",\"cell_cat\"]\n",
    "pheno_df.to_csv(prj_data_dir / 'metadata/pheno.tsv',sep=\"\\t\",index=False)\n",
    "\n",
    "metadata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>biosample</th>\n",
       "      <th>sra</th>\n",
       "      <th>geo</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>cell_cat</th>\n",
       "      <th>donor_cat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Donor 36 CD14+ mono</td>\n",
       "      <td>SAMN03785028</td>\n",
       "      <td>SRS967589</td>\n",
       "      <td>GSM1717162</td>\n",
       "      <td>SRR2072610</td>\n",
       "      <td>CD14+ mono</td>\n",
       "      <td>Donor 36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Donor 36 SLAN DC</td>\n",
       "      <td>SAMN03785027</td>\n",
       "      <td>SRS967590</td>\n",
       "      <td>GSM1717161</td>\n",
       "      <td>SRR2072609</td>\n",
       "      <td>SLAN DC</td>\n",
       "      <td>Donor 36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Donor 36 CD1c+ mDC</td>\n",
       "      <td>SAMN03785026</td>\n",
       "      <td>SRS967591</td>\n",
       "      <td>GSM1717160</td>\n",
       "      <td>SRR2072608</td>\n",
       "      <td>CD1c+ mDC</td>\n",
       "      <td>Donor 36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Donor 36 pDC</td>\n",
       "      <td>SAMN03785025</td>\n",
       "      <td>SRS967593</td>\n",
       "      <td>GSM1717159</td>\n",
       "      <td>SRR2072607</td>\n",
       "      <td>pDC</td>\n",
       "      <td>Donor 36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Donor 28 CD14+ Mono</td>\n",
       "      <td>SAMN03785024</td>\n",
       "      <td>SRS967592</td>\n",
       "      <td>GSM1717158</td>\n",
       "      <td>SRR2072606</td>\n",
       "      <td>CD14+ Mono</td>\n",
       "      <td>Donor 28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                sample     biosample        sra         geo run_accession  \\\n",
       "0  Donor 36 CD14+ mono  SAMN03785028  SRS967589  GSM1717162    SRR2072610   \n",
       "1     Donor 36 SLAN DC  SAMN03785027  SRS967590  GSM1717161    SRR2072609   \n",
       "2   Donor 36 CD1c+ mDC  SAMN03785026  SRS967591  GSM1717160    SRR2072608   \n",
       "3         Donor 36 pDC  SAMN03785025  SRS967593  GSM1717159    SRR2072607   \n",
       "4  Donor 28 CD14+ Mono  SAMN03785024  SRS967592  GSM1717158    SRR2072606   \n",
       "\n",
       "     cell_cat donor_cat  \n",
       "0  CD14+ mono  Donor 36  \n",
       "1     SLAN DC  Donor 36  \n",
       "2   CD1c+ mDC  Donor 36  \n",
       "3         pDC  Donor 36  \n",
       "4  CD14+ Mono  Donor 28  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pheno_fname = prj_data_dir / 'metadata/pheno.tsv'\n",
    "metadata_df = pd.read_csv(prj_data_dir / 'metadata/metadata.tsv', sep='\\t')\n",
    "metadata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd_lst_fname = prj_fastq_dir / 'cmd_lst.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"run_accession\"].tolist():\n",
    "        os.makedirs(prj_fastq_dir, exist_ok=True)\n",
    "        cmd_dump = f\"fasterq-dump {run_id} --outdir {prj_fastq_dir} --split-3\"\n",
    "        outFP.write(f\"{cmd_dump}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# align samples with hisat\n",
    "cmd_lst_fname = prj_aln_dir / 'cmd_lst.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"run_accession\"].tolist():\n",
    "        cmd = f\"hisat2 -p 25 --score-min L,0,-2 --mp 2,2 -x {hisat_idx_basename} -U {prj_fastq_dir}/{run_id}.fastq -S {prj_aln_dir}/{run_id}.sam\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sort reads by position\n",
    "cmd_lst_fname = prj_aln_dir / 'cmd_lst.sort.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"run_accession\"].tolist():\n",
    "        cmd = f\"samtools sort -@ 25 -o {prj_aln_dir}/{run_id}.sorted.bam {prj_aln_dir}/{run_id}.sam\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert to transcriptomic\n",
    "cmd_lst_fname = prj_aln_dir / 'cmd_lst.mudskipper.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"run_accession\"].tolist():\n",
    "        cmd = f\"{soft_dir / 'mudskipper/target/release/mudskipper'} bulk --shuffle --threads 25 --index {mudskipper_idx_basename} --alignment {prj_aln_dir}/{run_id}.sorted.bam --out {prj_aln_dir}/{run_id}.transcriptome.bam\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assemble with stringtie\n",
    "cmd_lst_fname = prj_assembly_dir / 'cmd_lst.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"run_accession\"].tolist():\n",
    "        cmd = f\"stringtie -p 25 -G {reference_gtf_fname} -o {prj_assembly_dir}/{run_id}.gtf {prj_aln_dir}/{run_id}.sorted.bam\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# quantify with salmon\n",
    "cmd_lst_fname = prj_quant_dir / 'cmd_lst.salmon.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"run_accession\"].tolist():\n",
    "        cmd = f\"salmon quant -i {salmon_idx_basename} -l A -1 {prj_fastq_dir}/{run_id}_1.fastq -2 {prj_fastq_dir}/{run_id}_2.fastq -p 25 -o {prj_quant_dir}/{run_id}\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>SRR2072610</th>\n",
       "      <th>SRR2072609</th>\n",
       "      <th>SRR2072608</th>\n",
       "      <th>SRR2072607</th>\n",
       "      <th>SRR2072606</th>\n",
       "      <th>SRR2072605</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.1.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.100.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.100.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.100.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.100.8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  transcript_id  SRR2072610  SRR2072609  SRR2072608  SRR2072607  SRR2072606  \\\n",
       "0       CHS.1.1           0           0           0           0           0   \n",
       "1     CHS.100.1           0           0           0           0           0   \n",
       "2     CHS.100.2           0           0           0           0           0   \n",
       "3     CHS.100.3           0           0           0           0           0   \n",
       "4     CHS.100.8           0           0           0           0           0   \n",
       "\n",
       "   SRR2072605  \n",
       "0           0  \n",
       "1           0  \n",
       "2           0  \n",
       "3           0  \n",
       "4           0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# generate counts matrix by merging counts files for all samples\n",
    "counts_fname = prj_quant_dir / 'counts_matrix.tsv'\n",
    "counts_df = pd.DataFrame(columns=['transcript_id'])\n",
    "for run_id in metadata_df[\"run_accession\"].tolist():\n",
    "    if not (prj_quant_dir / f\"{run_id}.counts\").exists():\n",
    "        continue\n",
    "    tmp_df = pd.read_csv(prj_quant_dir / f\"{run_id}.counts\", sep='\\t', header=None, names=['transcript_id', run_id])\n",
    "    if len(tmp_df) == 0:\n",
    "        continue\n",
    "    counts_df = pd.merge(counts_df, tmp_df, on='transcript_id', how='outer')\n",
    "counts_df.to_csv(counts_fname, sep='\\t', index=False)\n",
    "\n",
    "# create normalized matrix with cpm values\n",
    "cpm_fname = prj_quant_dir / 'cpm_matrix.tsv'\n",
    "cpm_df = counts_df.copy()\n",
    "cpm_df = cpm_df.set_index('transcript_id')\n",
    "cpm_df = cpm_df.div(cpm_df.sum(axis=0), axis=1) * 1e6\n",
    "cpm_df.reset_index(inplace=True)\n",
    "cpm_df.to_csv(cpm_fname, sep='\\t', index=False)\n",
    "cpm_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run spit\n",
    "spit_dir = prj_quant_dir / 'spit'\n",
    "spit_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "cmd = f\"spit preprocess -i {cpm_fname} -l {pheno_fname} -m {t2g_fname} -O {spit_dir} --write\"\n",
    "print(cmd)\n",
    "subprocess.run(cmd, shell=True)\n",
    "\n",
    "cmd = f\"spit dtu -i {cpm_fname} -l {pheno_fname} -m {t2g_fname} -O {spit_dir} --n_iter 10 --plot\"\n",
    "print(cmd)\n",
    "subprocess.run(cmd, shell=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hivAtlas",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
