{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Functional impairment of HIV-specific CD8+ T cells precedes aborted spontaneous control of viremia"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "project page: https://www.ncbi.nlm.nih.gov/bioproject/?term=HIV+longitudinal+RNA-Seq\n",
    "study: https://www.cell.com/immunity/fulltext/S1074-7613(21)00337-X?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS107476132100337X%3Fshowall%3Dtrue\n",
    "biosample_results: https://www.ncbi.nlm.nih.gov/biosample?Db=biosample&DbFrom=bioproject&Cmd=Link&LinkName=bioproject_biosample&LinkReadableName=BioSample&ordinalpos=1&IdsFromResult=706759"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/ccb/salz8-3/avaraby1/HIV_Atlas/.conda/lib/python3.11/site-packages/pysradb/utils.py:14: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from tqdm.autonotebook import tqdm\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import time\n",
    "import random\n",
    "import requests\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "from typing import List, Union\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from pysradb import SRAweb\n",
    "\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = Path.cwd().parent.parent.parent / 'HIV_Atlas_Creation' / 'data'\n",
    "\n",
    "sequence_dir = data_dir / 'sequences'\n",
    "assert sequence_dir.exists(), f\"sequence_dir does not exist: {sequence_dir}\"\n",
    "\n",
    "annotation_dir = data_dir / 'annotation'\n",
    "assert annotation_dir.exists(), f\"annotation_dir does not exist: {annotation_dir}\"\n",
    "\n",
    "reference_fasta_fname = sequence_dir / 'K03454.fasta'\n",
    "reference_gtf_fname = annotation_dir / 'K03454/K03454.vira.gtf'\n",
    "\n",
    "base_dir = Path.cwd().parent.parent\n",
    "\n",
    "prj_id = 'PRJNA706759'\n",
    "\n",
    "outdir = base_dir / 'results' / prj_id\n",
    "outdir.mkdir(parents=True, exist_ok=True)\n",
    "prj_data_dir = outdir / 'data'\n",
    "prj_data_dir.mkdir(parents=True, exist_ok=True)\n",
    "prj_fastq_dir = prj_data_dir / 'fastq'\n",
    "prj_fastq_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "prj_aln_dir = outdir / 'alignment'\n",
    "prj_aln_dir.mkdir(parents=True, exist_ok=True)\n",
    "prj_assembly_dir = outdir / 'assembly'\n",
    "prj_assembly_dir.mkdir(parents=True, exist_ok=True)\n",
    "prj_quant_dir = outdir / 'quantification'\n",
    "prj_quant_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "hisat_idx_basename = base_dir / 'data' / 'hg38_K03454'\n",
    "reference_fasta_fname = base_dir / 'data' / 'hg38_K03454.fasta'\n",
    "reference_gtf_fname = base_dir / 'data' / 'hg38_K03454.gtf'\n",
    "\n",
    "biosample_results_fname = prj_data_dir / 'metadata/biosample_results.tsv'\n",
    "sample_metadata_fname = prj_data_dir / 'metadata/sample_metadata.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load metadata\n",
    "biosample_results_df = pd.read_csv(biosample_results_fname, sep='\\t')\n",
    "sample_metadata_df = pd.read_csv(sample_metadata_fname, sep='\\t')\n",
    "metadata_df = biosample_results_df.merge(sample_metadata_df, left_on='sample', right_on=\"title\")\n",
    "metadata_df = metadata_df.drop(columns=['title'])\n",
    "\n",
    "db = SRAweb()\n",
    "batch_results = db.sra_metadata(metadata_df[\"sra\"].tolist(), detailed=True)\n",
    "metadata_df = metadata_df.merge(batch_results[[\"run_accession\",\"sample_accession\"]],left_on=\"sra\",right_on=\"sample_accession\",how=\"left\")\n",
    "metadata_df = metadata_df.drop(columns=[\"sample_accession\"])\n",
    "metadata_df.to_csv(prj_data_dir / 'metadata/metadata.tsv',sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>biosample</th>\n",
       "      <th>sra</th>\n",
       "      <th>geo</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Sample name</th>\n",
       "      <th>source name</th>\n",
       "      <th>organism</th>\n",
       "      <th>characteristics: PatientID</th>\n",
       "      <th>characteristics: Patient.Category</th>\n",
       "      <th>...</th>\n",
       "      <th>characteristics: Total.Reads</th>\n",
       "      <th>characteristics: Total.Align</th>\n",
       "      <th>processed data file</th>\n",
       "      <th>FileName.R1</th>\n",
       "      <th>FileName.R2</th>\n",
       "      <th>UnalignableReads</th>\n",
       "      <th>UniqueAlignReads</th>\n",
       "      <th>MultipleMapReads</th>\n",
       "      <th>MultiLocReads</th>\n",
       "      <th>run_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DC04 at T3 2</td>\n",
       "      <td>SAMN18143472</td>\n",
       "      <td>SRS8382647</td>\n",
       "      <td>GSM5136118</td>\n",
       "      <td>30</td>\n",
       "      <td>A0173.S0039</td>\n",
       "      <td>Sorted tetramer positive CD8+ lymphocytes</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DC04</td>\n",
       "      <td>DC</td>\n",
       "      <td>...</td>\n",
       "      <td>56654264</td>\n",
       "      <td>179668814</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A0173.S0039_R1_001.fastq.gz</td>\n",
       "      <td>A0173.S0039_R2_001.fastq.gz</td>\n",
       "      <td>25862430</td>\n",
       "      <td>17946316</td>\n",
       "      <td>12845342</td>\n",
       "      <td>24312056</td>\n",
       "      <td>SRR13860301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DC03 at T1 1</td>\n",
       "      <td>SAMN18143471</td>\n",
       "      <td>SRS8382648</td>\n",
       "      <td>GSM5136119</td>\n",
       "      <td>31</td>\n",
       "      <td>A0173.S0040</td>\n",
       "      <td>Sorted tetramer positive CD8+ lymphocytes</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DC03</td>\n",
       "      <td>DC</td>\n",
       "      <td>...</td>\n",
       "      <td>27435242</td>\n",
       "      <td>29278363</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A0173.S0040_R1_001.fastq.gz</td>\n",
       "      <td>A0173.S0040_R2_001.fastq.gz</td>\n",
       "      <td>21991807</td>\n",
       "      <td>2787066</td>\n",
       "      <td>2656360</td>\n",
       "      <td>4476084</td>\n",
       "      <td>SRR13860302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DC03 at T1 2</td>\n",
       "      <td>SAMN18143470</td>\n",
       "      <td>SRS8382649</td>\n",
       "      <td>GSM5136120</td>\n",
       "      <td>32</td>\n",
       "      <td>A0173.S0041</td>\n",
       "      <td>Sorted tetramer positive CD8+ lymphocytes</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DC03</td>\n",
       "      <td>DC</td>\n",
       "      <td>...</td>\n",
       "      <td>36528743</td>\n",
       "      <td>46938417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A0173.S0041_R1_001.fastq.gz</td>\n",
       "      <td>A0173.S0041_R2_001.fastq.gz</td>\n",
       "      <td>27570111</td>\n",
       "      <td>4985675</td>\n",
       "      <td>3972951</td>\n",
       "      <td>7225382</td>\n",
       "      <td>SRR13860303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DC03 at T1 3</td>\n",
       "      <td>SAMN18143469</td>\n",
       "      <td>SRS8382650</td>\n",
       "      <td>GSM5136121</td>\n",
       "      <td>33</td>\n",
       "      <td>A0173.S0042</td>\n",
       "      <td>Sorted tetramer positive CD8+ lymphocytes</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DC03</td>\n",
       "      <td>DC</td>\n",
       "      <td>...</td>\n",
       "      <td>34838174</td>\n",
       "      <td>47823066</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A0173.S0042_R1_001.fastq.gz</td>\n",
       "      <td>A0173.S0042_R2_001.fastq.gz</td>\n",
       "      <td>25543489</td>\n",
       "      <td>5303330</td>\n",
       "      <td>3991353</td>\n",
       "      <td>7451740</td>\n",
       "      <td>SRR13860304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DC03 at T3 3</td>\n",
       "      <td>SAMN18143468</td>\n",
       "      <td>SRS8382656</td>\n",
       "      <td>GSM5136127</td>\n",
       "      <td>39</td>\n",
       "      <td>A0173.S0048</td>\n",
       "      <td>Sorted tetramer positive CD8+ lymphocytes</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DC03</td>\n",
       "      <td>DC</td>\n",
       "      <td>...</td>\n",
       "      <td>27339863</td>\n",
       "      <td>27651869</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A0173.S0048_R1_001.fastq.gz</td>\n",
       "      <td>A0173.S0048_R2_001.fastq.gz</td>\n",
       "      <td>21872951</td>\n",
       "      <td>3198246</td>\n",
       "      <td>2268663</td>\n",
       "      <td>4377243</td>\n",
       "      <td>SRR13860310</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sample     biosample         sra         geo  Unnamed: 0  \\\n",
       "0  DC04 at T3 2  SAMN18143472  SRS8382647  GSM5136118          30   \n",
       "1  DC03 at T1 1  SAMN18143471  SRS8382648  GSM5136119          31   \n",
       "2  DC03 at T1 2  SAMN18143470  SRS8382649  GSM5136120          32   \n",
       "3  DC03 at T1 3  SAMN18143469  SRS8382650  GSM5136121          33   \n",
       "4  DC03 at T3 3  SAMN18143468  SRS8382656  GSM5136127          39   \n",
       "\n",
       "   Sample name                                source name      organism  \\\n",
       "0  A0173.S0039  Sorted tetramer positive CD8+ lymphocytes  Homo sapiens   \n",
       "1  A0173.S0040  Sorted tetramer positive CD8+ lymphocytes  Homo sapiens   \n",
       "2  A0173.S0041  Sorted tetramer positive CD8+ lymphocytes  Homo sapiens   \n",
       "3  A0173.S0042  Sorted tetramer positive CD8+ lymphocytes  Homo sapiens   \n",
       "4  A0173.S0048  Sorted tetramer positive CD8+ lymphocytes  Homo sapiens   \n",
       "\n",
       "  characteristics: PatientID characteristics: Patient.Category  ...  \\\n",
       "0                       DC04                                DC  ...   \n",
       "1                       DC03                                DC  ...   \n",
       "2                       DC03                                DC  ...   \n",
       "3                       DC03                                DC  ...   \n",
       "4                       DC03                                DC  ...   \n",
       "\n",
       "  characteristics: Total.Reads  characteristics: Total.Align  \\\n",
       "0                     56654264                     179668814   \n",
       "1                     27435242                      29278363   \n",
       "2                     36528743                      46938417   \n",
       "3                     34838174                      47823066   \n",
       "4                     27339863                      27651869   \n",
       "\n",
       "  processed data file                  FileName.R1  \\\n",
       "0                 NaN  A0173.S0039_R1_001.fastq.gz   \n",
       "1                 NaN  A0173.S0040_R1_001.fastq.gz   \n",
       "2                 NaN  A0173.S0041_R1_001.fastq.gz   \n",
       "3                 NaN  A0173.S0042_R1_001.fastq.gz   \n",
       "4                 NaN  A0173.S0048_R1_001.fastq.gz   \n",
       "\n",
       "                   FileName.R2  UnalignableReads  UniqueAlignReads  \\\n",
       "0  A0173.S0039_R2_001.fastq.gz          25862430          17946316   \n",
       "1  A0173.S0040_R2_001.fastq.gz          21991807           2787066   \n",
       "2  A0173.S0041_R2_001.fastq.gz          27570111           4985675   \n",
       "3  A0173.S0042_R2_001.fastq.gz          25543489           5303330   \n",
       "4  A0173.S0048_R2_001.fastq.gz          21872951           3198246   \n",
       "\n",
       "   MultipleMapReads  MultiLocReads  run_accession  \n",
       "0          12845342       24312056    SRR13860301  \n",
       "1           2656360        4476084    SRR13860302  \n",
       "2           3972951        7225382    SRR13860303  \n",
       "3           3991353        7451740    SRR13860304  \n",
       "4           2268663        4377243    SRR13860310  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_df = pd.read_csv(prj_data_dir / 'metadata/metadata.tsv', sep='\\t')\n",
    "metadata_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd_lst_fname = prj_fastq_dir / 'cmd_lst.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"run_accession\"].tolist():\n",
    "        os.makedirs(prj_fastq_dir, exist_ok=True)\n",
    "        cmd_dump = f\"fasterq-dump {run_id} --outdir {prj_fastq_dir} --split-3\"\n",
    "        outFP.write(f\"{cmd_dump}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# align samples with hisat\n",
    "cmd_lst_fname = prj_aln_dir / 'cmd_lst.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"run_accession\"].tolist():\n",
    "        cmd = f\"hisat2 -p 25 --score-min L,0,-2 --mp 2,2 -x {hisat_idx_basename} -1 {prj_fastq_dir}/{run_id}_1.fastq -2 {prj_fastq_dir}/{run_id}_2.fastq -S {prj_aln_dir}/{run_id}.sam\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# align samples with hisat\n",
    "cmd_lst_fname = prj_aln_dir / 'cmd_lst.sort.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"run_accession\"].tolist():\n",
    "        cmd = f\"samtools sort -@ 25 -o {prj_aln_dir}/{run_id}.sorted.bam {prj_aln_dir}/{run_id}.sam\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assemble with stringtie\n",
    "cmd_lst_fname = prj_assembly_dir / 'cmd_lst.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"run_accession\"].tolist():\n",
    "        cmd = f\"stringtie -p 25 -G {reference_gtf_fname} -o {prj_assembly_dir}/{run_id}.gtf {prj_aln_dir}/{run_id}.sorted.bam\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run htseq-count to quantify transcript expression\n",
    "cmd_lst_fname = prj_quant_dir / 'cmd_lst.txt'\n",
    "with open(cmd_lst_fname, 'w') as outFP:\n",
    "    for run_id in metadata_df[\"run_accession\"].tolist():\n",
    "        cmd = f\"htseq-count -n 20 -f bam -s no -t exon -i transcript_id {prj_aln_dir}/{run_id}.sorted.bam {reference_gtf_fname} > {prj_quant_dir}/{run_id}.counts\"\n",
    "        outFP.write(f\"{cmd}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
