{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import time\n",
    "import shutil\n",
    "import random\n",
    "import sqlite3\n",
    "import requests\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accid</th>\n",
       "      <th>organism</th>\n",
       "      <th>status</th>\n",
       "      <th>genome_score</th>\n",
       "      <th>junction_score</th>\n",
       "      <th>transcript_score</th>\n",
       "      <th>miniprot_score</th>\n",
       "      <th>guide_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MH705157.1</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>complete</td>\n",
       "      <td>0.871767</td>\n",
       "      <td>0.944444</td>\n",
       "      <td>0.799090</td>\n",
       "      <td>0.661600</td>\n",
       "      <td>0.559152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DQ396400.1</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>complete</td>\n",
       "      <td>0.901976</td>\n",
       "      <td>0.958333</td>\n",
       "      <td>0.845619</td>\n",
       "      <td>0.661978</td>\n",
       "      <td>0.720496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KU168256.1</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>complete</td>\n",
       "      <td>0.878658</td>\n",
       "      <td>0.937500</td>\n",
       "      <td>0.819815</td>\n",
       "      <td>0.559410</td>\n",
       "      <td>0.719850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MH705158.1</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>complete</td>\n",
       "      <td>0.874285</td>\n",
       "      <td>0.958333</td>\n",
       "      <td>0.790237</td>\n",
       "      <td>0.407212</td>\n",
       "      <td>0.753737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH705151.1</td>\n",
       "      <td>HIV-1</td>\n",
       "      <td>complete</td>\n",
       "      <td>0.823610</td>\n",
       "      <td>0.914352</td>\n",
       "      <td>0.732868</td>\n",
       "      <td>0.391839</td>\n",
       "      <td>0.564241</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accid organism    status  genome_score  junction_score  \\\n",
       "0  MH705157.1    HIV-1  complete      0.871767        0.944444   \n",
       "1  DQ396400.1    HIV-1  complete      0.901976        0.958333   \n",
       "2  KU168256.1    HIV-1  complete      0.878658        0.937500   \n",
       "3  MH705158.1    HIV-1  complete      0.874285        0.958333   \n",
       "4  MH705151.1    HIV-1  complete      0.823610        0.914352   \n",
       "\n",
       "   transcript_score  miniprot_score  guide_score  \n",
       "0          0.799090        0.661600     0.559152  \n",
       "1          0.845619        0.661978     0.720496  \n",
       "2          0.819815        0.559410     0.719850  \n",
       "3          0.790237        0.407212     0.753737  \n",
       "4          0.732868        0.391839     0.564241  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "base_dir = Path.cwd().parent.parent\n",
    "data_dir = base_dir / 'HIV_Atlas_Creation/data'\n",
    "\n",
    "reference_data = {\n",
    "    11676: {\"accid\": \"K03455.1\",\n",
    "            \"organism\": \"HIV-1\",\n",
    "            \"fasta\": data_dir / \"reference/K03455.1.fasta\",\n",
    "            \"gtf\": data_dir / \"reference/K03455.1.gtf\",\n",
    "            \"description\": 'Human immunodeficiency virus type 1 (HXB2), complete genome; HIV1/HTLV-III/LAV reference genome'},\n",
    "    11723: {\"accid\": \"M33262.1\",\n",
    "            \"organism\": \"SIV\",\n",
    "            \"fasta\": data_dir / \"reference/M33262.1.fasta\",\n",
    "            \"gtf\": data_dir / \"reference/M33262.1.gtf\",\n",
    "            \"description\": 'Simian (macaque) immunodeficiency virus, isolate 239, complete proviral genome.'}\n",
    "}\n",
    "reference_ids = set([x[\"accid\"] for x in reference_data.values()])\n",
    "\n",
    "sequence_dir = data_dir / 'sequences'\n",
    "assert sequence_dir.exists(), f\"sequence_dir does not exist: {sequence_dir}\"\n",
    "\n",
    "annotation_dir = data_dir / 'annotation'\n",
    "assert annotation_dir.exists(), f\"annotation_dir does not exist: {annotation_dir}\"\n",
    "\n",
    "outdir = data_dir / 'conservation'\n",
    "outdir.mkdir(exist_ok=True)\n",
    "\n",
    "accession_list_fname = data_dir / 'complete_sequences.2022.accessions'\n",
    "summary_file =data_dir / \"download.summary.tsv\"\n",
    "\n",
    "annotation_log_fname = data_dir / 'annotation/log.txt'\n",
    "# load the annotation log file to extract relevant data into the database tsv\n",
    "log_df = pd.read_csv(annotation_log_fname)\n",
    "\n",
    "complete_log_df = log_df[log_df[\"status\"] == \"complete\"].reset_index(drop=True)\n",
    "complete_log_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the junction information\n",
    "\n",
    "# for every donor/acceptor site on the reference genome\n",
    "# we should load counts of every nucleotide around it\n",
    "# load list of genomes\n",
    "\n",
    "def load_junctions(junction_stats_fname,sites):\n",
    "    # loads data from the junction_stats_fname file and organizes it into the sites dictionary\n",
    "    with open(junction_stats_fname, \"r\") as inFP:\n",
    "        reader = csv.reader(inFP, delimiter=\"\\t\")\n",
    "        # skip header\n",
    "        next(reader)\n",
    "        for lno, (position, query_position, site_type, map_consistency, sequence, query_sequence) in enumerate(reader):\n",
    "            query_position = int(query_position)\n",
    "            # sequence is offset+2bp site+offset\n",
    "            assert len(sequence)%2==0, f\"Error: sequence length is not even: {len(sequence)}\"\n",
    "            offset = int((len(sequence)-2)/2)\n",
    "            start_pos = query_position - offset\n",
    "            for pos in range(start_pos, start_pos+len(sequence)):\n",
    "                sites[site_type].setdefault(pos, {\"A\":0,\"C\":0,\"G\":0,\"T\":0,\"N\":0})\n",
    "                nt = sequence[pos-start_pos]\n",
    "                if not nt in \"ACGT\":\n",
    "                    nt = \"N\"\n",
    "                sites[site_type][pos][nt] += 1\n",
    "            \n",
    "\n",
    "sites = {\"acceptor\":{},\n",
    "         \"donor\":{}}\n",
    "with open(summary_file,\"r\") as inFP:\n",
    "    reader = csv.reader(inFP, delimiter=\"\\t\")\n",
    "    # skip header\n",
    "    next(reader)\n",
    "    for lno, (accid, is_ref_genome, organism, taxid, fasta, gff, is_valid, ref_start, ref_end) in enumerate(reader):\n",
    "        if accid in reference_ids:\n",
    "            continue\n",
    "\n",
    "        if (annotation_dir / accid / f\"{accid}.vira.gtf\").exists():\n",
    "            # load line from the annotation log to get the score\n",
    "            if len(log_df[log_df[\"accid\"]==accid]) == 0:\n",
    "                print(f\"Warning: {accid} not found in log file, skipping\")\n",
    "                continue\n",
    "            \n",
    "            log_row = log_df[log_df[\"accid\"]==accid]\n",
    "            assert len(log_row) == 1, f\"Error: multiple rows found for {accid}\"\n",
    "\n",
    "            status = log_row[\"status\"].values[0]\n",
    "            if not status == \"complete\":\n",
    "                print(f\"Warning: {accid} is not complete, skipping\")\n",
    "                continue\n",
    "            \n",
    "            if not int(taxid) == 11676:\n",
    "                continue\n",
    "            \n",
    "            acc_outdir = annotation_dir / accid\n",
    "            acc_outdir.mkdir(exist_ok=True)\n",
    "            junction_stats_fname = acc_outdir / f\"{accid}.vira.junction_stats.tsv\"\n",
    "            \n",
    "            load_junctions(junction_stats_fname, sites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write this information out to files (donors and acceptors separately perhaps)\n",
    "for site_type in sites:\n",
    "    outfname = outdir / f\"junctions_{site_type}.tsv\"\n",
    "    with open(outfname, \"w\") as outFP:\n",
    "        outFP.write(\"seqid\\tposition\\tA\\tC\\tG\\tT\\tN\\n\")\n",
    "        for pos in sorted(sites[site_type]):\n",
    "            outFP.write(f\"{reference_data[11676]['accid']}\\t{pos}\")\n",
    "            for nt in \"ACGTN\":\n",
    "                outFP.write(f\"\\t{sites[site_type][pos][nt]}\")\n",
    "            outFP.write(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hivAtlas",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
